pkg/prototype/values_test.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2018 The ksonnet authors
+//
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package prototype
+
+import (
+	"testing"
+
+	ksstrings "github.com/ksonnet/ksonnet/pkg/util/strings"
+	"github.com/ksonnet/ksonnet/pkg/util/test"
+	"github.com/spf13/afero"
+	"github.com/spf13/pflag"
+	"github.com/stretchr/testify/require"
+)
+
+func TestValues(t *testing.T) {
+	proto := &Prototype{
+		Params: ParamSchemas{
+			{
+				Name: "name",
+				Type: String,
+			},
+			{
+				Name: "key1",
+				Type: String,
+			},
+			{
+				Name:    "key2",
+				Type:    String,
+				Default: ksstrings.Ptr("default"),
+			},
+		},
+	}
+
+	cases := []struct {
+		name     string
+		p        *Prototype
+		flags    *pflag.FlagSet
+		args     []string
+		isErr    bool
+		expected map[string]string
+	}{
+		{
+			name: "from flags",
+			p:    proto,
+			args: []string{
+				"--name", "name",
+			},
+			flags: BindFlags(proto),
+			expected: map[string]string{
+				"name": `"name"`,
+				"key1": `""`,
+				"key2": `"default"`,
+			},
+		},
+		{
+			name:  "flags: prototype is nil",
+			flags: BindFlags(proto),
+			args:  []string{},
+			isErr: true,
+		},
+		{
+			name:  "flags: missing default parameter",
+			flags: BindFlags(proto),
+			args:  []string{},
+			isErr: true,
+		},
+		{
+			name:  "from values file",
+			p:     proto,
+			flags: BindFlags(proto),
+			args: []string{
+				"--name", "name",
+				"--values-file", "/files/valid",
+			},
+			expected: map[string]string{
+				"name": `"name"`,
+				"key1": `"value1"`,
+				"key2": `"default"`,
+			},
+		},
+		{
+			name:  "values: invalid",
+			p:     proto,
+			flags: BindFlags(proto),
+			args: []string{
+				"--name", "name",
+				"--values-file", "/files/invalid",
```

Comment:
If `=` is prohibited in values, can we add a test case capturing that?

Reply:
It's not prohibited, but the string is split into two parts. I'll add a test case for this.

Reply:
It looks like we're explicitly checking for a two-way split, which would preclude additional `=`:

```
		parts := strings.SplitN(s, "=", 2)
		if len(parts) != 2 {
			return nil, errors.Errorf("%q is invalid. value format is key=value", s)
		}
```

Reply:
No, that's not how it would work. Here's an example:

https://play.golang.org/p/_uR5RufO667

Reply:
Ah, so that is the difference between `Split` and `SplitN` 🙂
I misunderstood.